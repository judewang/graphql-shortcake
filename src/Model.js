/* eslint no-param-reassign: ["error", { "ignorePropertyModificationsFor": ["target"] }] */
import _ from 'lodash';
import fp from 'lodash/fp';
import { invokeNull } from 'thelper';
import withArrayMutator from './withArrayMutator';
import withBuilder from './withBuilder';
import withBatch from './withBatch';
import withCache from './withCache';
import withColumn from './withColumn';
import withDefaultValues from './withDefaultValues';
import withFetcher from './withFetcher';
import withGlobalId from './withGlobalId';
import withIncrementer from './withIncrementer';
import withJSONMutator from './withJSONMutator';
import withLoader from './withLoader';
import withMutator from './withMutator';
import withSearcher from './withSearcher';

class Model {
  static database = null;

  static tableName = null;

  static viewName = null;

  static idAttribute = 'id';

  static hasOperator = false;

  static hasTimestamps = true;

  static softDelete = true;

  static get(target, name) {
    if (this.has(target, name)) {
      return target.get(name);
    }

    return target[name];
  }

  static set(target, name, value) {
    if (this.has(target, name)) {
      target.set(name, value);
      return;
    }

    target[name] = value;
  }

  static get displayName() {
    return this.name;
  }

  static has(target, name) {
    return target.ownKeys.indexOf(name) > -1;
  }

  static ownKeys(target) {
    return target.ownKeys;
  }

  static getOwnPropertyDescriptor(target, name) {
    if (this.has(target, name)) {
      return { configurable: true, enumerable: true, value: target.get(name) };
    }

    return Object.getOwnPropertyDescriptor(target, name);
  }

  static refreshView = _.debounce(() => {
    this.raw(`REFRESH MATERIALIZED VIEW ${this.viewName};`);
  }, 500, { maxWait: 2000 });

  static raw(...args) {
    return this.database.raw(...args);
  }

  static forge(attributes, options) {
    const model = new this({}, options);
    return model.forge(attributes);
  }

  static fromModel(model) {
    return _.get(model, ['nativeId'], model);
  }

  static checkOperator(operator) {
    const { hasOperator } = this;
    if (!hasOperator) return null;
    if (!operator) throw new Error('operator is required');
    return this.fromModel(operator);
  }

  [Symbol.toStringTag] = 'Model';

  [Symbol.toPrimitive](hint) {
    if (hint === 'number') return Number.NaN;
    return this.toString();
  }

  constructor() {
    if (this.initialize) this.initialize();
    const handler = _.pick(this.constructor, [
      'get', 'set', 'has', 'ownKeys', 'getOwnPropertyDescriptor',
    ]);
    return new Proxy(this, handler);
  }

  initialize() {
    this._ = {};
    this.current = {};
    this.previous = {};
  }

  get isNew() {
    const { idAttribute } = this.constructor;
    return !this[idAttribute];
  }

  get values() {
    return { ...this.previous, ...this.current };
  }

  get(paths) {
    const { values } = this;
    return _.get(values, paths, null);
  }

  set(paths, value) {
    _.set(this.current, paths, value);
  }

  valueOf(name) {
    if (name) return _.get(this.current, [name], this.previous[name]);

    const { idAttribute } = this.constructor;
    const id = this.valueOf(idAttribute);
    return id ? String(id) : id;
  }

  toString() {
    const { idAttribute } = this.constructor;
    return this[idAttribute];
  }

  clone(options) {
    const model = new this.constructor({}, options);
    model.current = _.cloneDeep(this.current);
    model.previous = this.previous;
    return model;
  }

  forge(attributes) {
    return invokeNull((attrs) => {
      const values = _.get(attrs, ['previous'], attrs);
      this.current = _.cloneDeep(values);
      this.previous = values;
      return this;
    }, attributes);
  }

  merge(attributes) {
    return invokeNull((attrs) => {
      const { format } = this.constructor;
      const values = format(attrs);
      _.forEach([this.current, this.previous], data => _.assign(data, values));
      return this;
    }, attributes);
  }
}

export default fp.compose(
  withArrayMutator,
  withJSONMutator,
  withIncrementer,
  withSearcher,
  withDefaultValues,
  withCache,
  withLoader,
  withFetcher,
  withMutator,
  withBatch,
  withBuilder,
  withColumn,
  withGlobalId,
)(Model);
