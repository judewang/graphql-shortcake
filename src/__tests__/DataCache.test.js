import _ from 'lodash';
import knex, { client } from 'knex';
import Model, { DataCache, toGlobalId } from '..';

const database = knex({
  client: 'pg',
  connection: {
    host: '127.0.0.1',
    user: 'postgres',
    password: null,
    database: 'graphql_tower',
  },
});

class Car extends Model {
  static database = database;

  static tableName = 'car';

  static columns = {
    name: { type: String },
  }
}

class House extends Model {
  static database = database;

  static tableName = 'house';

  static columns = {
    name: { type: String },
  }
}

class Cache extends DataCache {
  static Car = Car;

  static House = House;
}

describe('DataCache', () => {
  afterAll(() => database.destroy());

  describe('load & loadMany', () => {
    it('cache 1', async () => {
      const cache = new Cache();

      client.mockReturnValueOnce([{ id: '1', name: '1 of car' }]);
      await cache.load(Car.toGlobalId('1'));

      client.mockReturnValueOnce([{ id: '1', name: '1 of house' }]);
      await cache.load(House.toGlobalId('1'));

      client.mockReturnValueOnce([{ id: '2', name: '2 of house' }, { id: '4', name: '4 of house' }]);
      client.mockReturnValueOnce([{ id: '3', name: '3 of car' }]);

      await Promise.all([
        cache.loadCar(Car.toGlobalId('1')),
        cache.loadHouse(House.toGlobalId('2')),
        cache.loadMany([Car.toGlobalId('3'), House.toGlobalId('1'), House.toGlobalId('4')]),
      ]);

      expect(client).toMatchSnapshot();
      expect(client).toHaveBeenCalledTimes(4);
    });

    it('cache 2', async () => {
      const cache = new Cache();

      client.mockReturnValueOnce([{ id: '1', name: '1 of car' }]);
      await cache.load(Car.toGlobalId('1'));

      client.mockReturnValueOnce([{ id: '1', name: '1 of house' }]);
      await cache.load(House.toGlobalId('1'));

      expect(client).toMatchSnapshot();
      expect(client).toHaveBeenCalledTimes(2);
    });
  });

  it('check type', async () => {
    const cache = new Cache();

    await expect(cache.load(toGlobalId('tree', '1'))).resolves.toBeNull();

    client.mockReturnValueOnce([]);
    await expect(cache.load(Car.toGlobalId('1')))
      .resolves.toBeNull();

    await expect(cache.load(Car.toGlobalId('1'), Error))
      .rejects.toEqual(new Error());

    await expect(cache.loadHouse(Car.toGlobalId('2'), TypeError))
      .rejects.toEqual(new TypeError());

    expect(client).toMatchSnapshot();
    expect(client).toHaveBeenCalledTimes(1);
  });

  describe('prime', async () => {
    it('when prime new data', async () => {
      const id = Car.toGlobalId('1');

      const cache = new Cache();

      const car = Car.forge({ id: '1', name: '1 of car' });
      expect(car.cache).toBeNull();
      cache.prime(id, car);
      expect(car.cache).toBe(cache);
      await expect(cache.load(id))
        .resolves.toEqual(expect.objectContaining({ id, name: '1 of car' }));
      expect(client).toHaveBeenCalledTimes(0);

      cache.prime(id);
      client.mockReturnValueOnce([{ id: '1', name: '1 of car again' }]);
      await expect(cache.load(id))
        .resolves.toEqual(expect.objectContaining({ id, name: '1 of car again' }));
      expect(client).toHaveBeenCalledTimes(1);
    });

    it('when model does not exist', () => {
      const cache = new Cache();
      cache.dataloader = jest.fn();
      cache.prime(toGlobalId('tree', '1'));

      expect(cache.dataloader).toHaveBeenCalledTimes(0);
    });
  });

  it('clear & clearAll', async () => {
    const cache = new Cache();

    client.mockReturnValueOnce([{ id: '1', name: '1 of car' }]);
    client.mockReturnValueOnce([{ id: '1', name: '1 of car' }]);
    client.mockReturnValueOnce([{ id: '1', name: '1 of car' }]);

    await cache.load(Car.toGlobalId('1'));
    await cache.load(Car.toGlobalId('1'));
    cache.clear(Car.toGlobalId('1'));
    await cache.load(Car.toGlobalId('1'));
    cache.clearAll();
    await cache.load(Car.toGlobalId('1'));

    expect(client).toMatchSnapshot();
    expect(client).toHaveBeenCalledTimes(3);
  });

  it('new model', async () => {
    const cache = new Cache();
    const { car } = cache;

    client.mockReturnValueOnce([{ id: '1', name: '1 of car' }, { id: '2', name: '2 of car' }]);
    expect(_.size(await car.fetchAll())).toBe(2);

    await cache.load(Car.toGlobalId('1'));

    expect(client).toMatchSnapshot();
    expect(client).toHaveBeenCalledTimes(1);
  });
});
