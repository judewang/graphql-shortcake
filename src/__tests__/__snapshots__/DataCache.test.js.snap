// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataCache check type 1`] = `
Array [
  "select * from car where id in (1) and deleted_at is null",
]
`;

exports[`DataCache clear & clearAll 1`] = `
Array [
  "select * from car where id in (1) and deleted_at is null",
  "select * from car where id in (1) and deleted_at is null",
  "select * from car where id in (1) and deleted_at is null",
]
`;

exports[`DataCache load & loadMany cache 1 1`] = `
Array [
  "select * from car where id in (1) and deleted_at is null",
  "select * from house where id in (1) and deleted_at is null",
  "select * from house where id in (2, 4) and deleted_at is null",
  "select * from car where id in (3) and deleted_at is null",
]
`;

exports[`DataCache load & loadMany cache 2 1`] = `
Array [
  "select * from car where id in (1) and deleted_at is null",
  "select * from house where id in (1) and deleted_at is null",
]
`;

exports[`DataCache new model 1`] = `
Array [
  "select *, count(*) OVER() AS total_count from car where deleted_at is null",
]
`;
